webrender/src/mask_cache.rs

```diff
@@ -8,158 +8,273 @@ use prim_store::{CLIP_DATA_GPU_SIZE, MASK_DATA_GPU_SIZE};
+                            let data = ClipData::from_clip_region(clip);
+                            PrimitiveStore::populate_clip_data(chunk, data);
+                            local_rect = local_rect.and_then(|r| r.intersection(&clip.rect));
+                            local_inner = local_inner.and_then(|r| clip.get_inner_rect()
+                                                                       .and_then(|ref inner| r.intersection(&inner)));
+                        }
                     }
                 }
-            };
-            self.local_rect = Some(local_rect.unwrap_or(LayerRect::zero()));
-            self.local_inner = local_inner;
+            }
+
+            // Work out the type of mask geometry we have, based on the
+            // list of clip sources above.
+            if clip_out {
+                // For clip-out, the mask rect is not known.
+                self.bounds = MaskBounds::None;
+            } else {
+                // local inner is only valid if there's a single clip (for now).
+                // This can be improved in the future, with some proper
+                // rectangle region handling.
+                if sources.len() > 1 {
+                    local_inner = None;
+                }
+
+                let local_rect = local_rect.unwrap_or(LayerRect::zero());
+
+                self.bounds = match local_inner {
+                    Some(local_inner) => {
+                        MaskBounds::OuterInner(Geometry::new(local_rect),
+                                               Geometry::new(local_inner))
+                    }
+                    None => {
+                        MaskBounds::Outer(Geometry::new(local_rect))
+                    }
+                };
+            }
         }
 
-        let transformed = TransformedRect::new(self.local_rect.as_ref().unwrap(),
-                                               &transform,
-                                               device_pixel_ratio);
-        self.outer_rect = transformed.bounding_rect;
-
-        self.inner_rect = if let Some(ref inner_rect) = self.local_inner {
-            let transformed = TransformedRect::new(inner_rect,
-                                                   &transform,
-                                                   device_pixel_ratio);
-            transformed.inner_rect
-        } else {
-            DeviceIntRect::new(self.outer_rect.origin, DeviceIntSize::zero())
+        // Update the device space bounding rects of the mask
+        // geometry.
+        match self.bounds {
+            MaskBounds::Unknown => unreachable!(),
+            MaskBounds::None => {}
+            MaskBounds::Outer(ref mut outer) => {
+                outer.update(transform, device_pixel_ratio);
+            }
+            MaskBounds::OuterInner(ref mut outer, ref mut inner) => {
+                outer.update(transform, device_pixel_ratio);
+                inner.update(transform, device_pixel_ratio);
```

Comment:
I don't think this is correct. The `update` here would compute the bounding box of the inner area, which is useless to us (and confusing). Not sure if we actually use it anywhere, but it doesn't appear that `inner` fits the whole `Geometry` struct concept.